src/main/java/com/google/devtools/build/lib/cmdline/PackageIdentifier.java

```diff
@@ -177,6 +178,65 @@ public String getCanonicalForm() {
     return repository.getCanonicalForm() + "//" + getPackageFragment();
   }
 
+  /**
+   * Returns an absolutely unambiguous canonical form for this package in label form. Parsing this
+   * string in any environment, even when subject to repository mapping, should identify the same
+   * package.
+   */
+  public String getUnambiguousCanonicalForm() {
+    return String.format("@@%s//%s", getRepository().getName(), getPackageFragment());
+  }
+
+  /**
+   * Returns a label representation for this package that is suitable for display. The returned
+   * string is as simple as possible while referencing the current package when parsed in the
+   * context of the main repository.
+   *
+   * @param mainOrOtherRepoMapping the {@link RepositoryMapping} of the main repository or, if that
+   *                               is not available to the caller, the {@link RepositoryMapping} of
```

Comment:
it looks like if the repo mapping provided is not of the main repo, then it doesn't matter at all what the repo mapping is, right? it could be for any non-main repo and the result would be the same. We might as well say "grab any repo mapping object and pass it here".

this makes me a bit uneasy about this API, though I don't have a better idea right now.

Reply:
I wasn't completely sure about that: Are "Bzlmod is enabled" and "for each `RepositoryMapping` instance `ownerRepo()` is non-`null`" equivalent? I was worried of e.g. the `RepositoryMapping`s of repos defined in `WORKSPACE` fallback mode to have `ownerRepo() == null` even though they should really result in canonical labels (or not?). It would be great if you could clarify that point.

Reply:
As you suspected, they're not equivalent, exactly because WORKSPACE repos still aren't subject to strict deps. That doesn't mean they don't go through repo mapping, just that if a local name is missing from the mapping, it'll be used as the canonical name instead. This actually means that line 224 is broken in certain cases.

Reply:
Since the API of `getDisplayForm` was confusing and I would have had to pass in the Bzlmod bit anyway, I am now passing the main repository's mapping into `Package`. This results in the ideal error message and makes it so that `getDisplayForm` can expect to be passed the main repository's rather than any mapping.

Reply:
Nice!